jobs-builder: add PR jobs for Fedora 35

wainersm opened the pull request:
This created the backbone to generate pull request and baseline jobs
with Jenkins Job Builder. Currently the yaml files are configured to
generate PR and baseline jobs for Fedora 35.

The following new jobs will be generated:

kata-containers-fedora-35-x86_64-CRIO_K8S-PR
kata-containers-fedora-35-x86_64-CRIO_K8S_MINIMAL-PR
kata-containers-main-fedora-35-x86_64-CRIO_K8S-baseline
kata-containers-main-fedora-35-x86_64-CRIO_K8S_MINIMAL-baseline
tests-fedora-35-x86_64-CRIO_K8S-PR
tests-fedora-35-x86_64-CRIO_K8S_MINIMAL-PR

Fixes #389
Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

wainersm commented:
@jodh-intel @fidencio this is a continuation of #406. I'm adding jobs for Fedora 35 but the most important thing are the two added templates which can be used to generate PR and baseline jobs. With this merged I expect that migrating the jobs ( https://github.com/kata-containers/ci/issues/407) to be straightforward.

@GabyCT @devimc mind to review this PR? I basically copied the properties and configuration of the existing Ubuntu jobs (either PR and baseline). The only change that I made (and I am not so sure) is the use of the `ci_entry_point.sh` script on both baseline and PR jobs, but I've seen that script being used only on baselines. Does it work fine with PR jobs too?

I already published two jobs on our Jenkins just for the sake of testing. You can have a look how they will end up generated:
http://jenkins.katacontainers.io/job/kata-containers-main-fedora-35-x86_64-CRIO_K8S-baseline
http://jenkins.katacontainers.io/job/tests-fedora-35-x86_64-CRIO_K8S-PR/

devimc commented on jobs-builder/jobs/include/ci_entrypoint.sh.inc:
```diff
@@ -0,0 +1,17 @@
+# with expected environment variables exported.
+#
+
+#!/bin/bash
```
this must be the first line

wainersm replied:
True. Done.

devimc commented on jobs-builder/jobs/include/ci_entrypoint.sh.inc:
```diff
@@ -0,0 +1,17 @@
+
+curl -OL https://raw.githubusercontent.com/kata-containers/tests/main/.ci/ci_entry_point.sh
+export DEBUG=true
+export CI_JOB="{ci_job}"
```
missing `$`

devimc commented on jobs-builder/jobs/pr.yaml:
```diff
@@ -0,0 +1,180 @@
+
+              export GOPATH=$WORKSPACE/go
+              export GOROOT="/usr/local/go"
+              export PATH="${{GOPATH}}/bin:/usr/local/go/bin:/usr/sbin:/usr/local/bin:${{PATH}}"
```
why double `{` here and only  one `{` in the following line ?

devimc reviewed (commented):
thanks @wainersm - I've left some comments

jodh-intel reviewed (approved):
`{l}-{g}-{t}-{m}{ }{!} { :-) }` <= Not a fork bomb, honest! ;)

wainersm commented:
> `{l}-{g}-{t}-{m}{ }{!} { :-) }` <= Not a fork bomb, honest! ;)

Hehehehhe thanks for reviewing @jodh-intel !

wainersm commented:
Updates to this PR:
- Moved the shebang to the first line (@devimc)
- I realized the baseline jobs would be running 3x per day. Changed the `timed` value to `'H 0 * * *'`

@jodh-intel @GabyCT maybe don't need to approve again.

jodh-intel commented on jobs-builder/jobs/include/ci_entrypoint.sh.inc:
```diff
@@ -0,0 +1,16 @@
+# with expected environment variables exported.
+#
+
+set -e
+set -x
```
Nit: Might be worth adding the usual header:

```bash
set -o errexit
set -o nounset
set -o pipefail

[ -n "$BASH_VERSION" ] && set -o errtrace
[ -n "${DEBUG:-}" ] && set -o xtrace
```

wainersm replied:
This script is just a thin wrapper to https://raw.githubusercontent.com/kata-containers/tests/main/.ci/ci_entry_point.sh , which in turn adds all those flags :)

jodh-intel commented on jobs-builder/jobs/pr.yaml:
```diff
@@ -0,0 +1,180 @@
+              - log-text: .*
+                operator: OR
+            script: |
+              #!/bin/bash
```
Again, why not add the standard shell header here too?

Random thought: I wonder if we need a `.ci/kata_header.sh` to make this easy to source and minimise duplication.

wainersm replied:
Here is yet another thin wrapper, this time to https://raw.githubusercontent.com/kata-containers/tests/main/.ci/teardown.sh . I just realized that unlike `ci_entry_point.sh`, the `teardown.sh` script doesn't set those shell properties. Maybe it is by design because we don't want it to fail early? (it collects logs; one or more commands might fail and it is ok?)

wainersm replied:
Regarding adding the `.ci/kata_header.sh`; then people like me will forget to source it :)

jodh-intel commented on jobs-builder/jobs/pr.yaml:
```diff
@@ -0,0 +1,180 @@
+      - CRIO_K8S
+      - CRIO_K8S_MINIMAL
+    maintainers:
+      - '@snir911'
+      - '@c3d'
```
ooi, are you both in the same TZ? If so, maybe we could add a few more users: I'd say minimum of three, but even better would be to add a single GH team here.

wainersm replied:
We agreed that baseline jobs should be owners (https://github.com/kata-containers/ci/issues/403). @snir911 and @c3d are going to be the owners of those Fedora 35  jobs. So it is fine.

jodh-intel replied:
What does having a user specified in `maintainers` here actually mean though? What if they are both unavailable? Is it simply for mail notification or similar?

wainersm replied:
I'm not sure I understood your first question. The `maintainers` variable is used on the description of the job, so on the Jenkins UI it will be listed the names on the jobs' description. An example here: http://jenkins.katacontainers.io/view/Daily%20baseline/job/kata-containers-main-fedora-35-x86_64-CRIO_K8S-baseline/

The idea is to have their names on the job's description for the case a developer need to speak with the owners (e.g. developer's PR cannot be merged because a job is failing and she/he notices the baseline has failed too). The owners are also listed in a table on the view [here](http://jenkins.katacontainers.io/view/Daily%20baseline/). Having their names on the job's description just increase visibility. Do you see any problem with that?

jodh-intel reviewed (approved):
Thanks @wainersm. A few nits but overall,

lgtm

fidencio commented:
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

snir911 reviewed (approved):
LGTM, thanks!
(for the record, as discussed with @wainersm) We just need to make sure it indeed override/replace the f35 jobs i created manually, and also once we see everything works as expected, remove the f32 jobs

snir911 commented:
adding do-not-merge until Wainer's ack to merge

wainersm commented:
@jodh-intel can I have you ack?

wainersm commented:
> LGTM, thanks! (for the record, as discussed with @wainersm) We just need to make sure it indeed override/replace the f35 jobs i created manually, and also once we see everything works as expected, remove the f32 jobs

I just renamed the Fedora 35 @snir911 created manually so on publishing they will be overwritten (and we keep the historic of executions).
